pytorch_lightning/loops/prediction_epoch_loop.py

```diff
@@ -0,0 +1,143 @@
+from collections import OrderedDict
+from typing import Any, Dict, Iterator, List, Optional, Tuple
+
+from deprecate import void
+
+from pytorch_lightning.loops.base import Loop
+from pytorch_lightning.overrides.distributed import IndexBatchSamplerWrapper
+from pytorch_lightning.utilities.warnings import WarningCache
+
+
+class PredictionEpochLoop(Loop):
+    """Loop performing prediction on arbitrary sequentially used dataloaders."""
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.return_predictions: bool = False
+        self.predictions: List[Any] = []
+        self.current_batch_indices: List[int] = []
+        self._dl_max_batches: Optional[int] = None
+        self._num_dataloaders: Optional[int] = None
+        self._warning_cache = WarningCache()
+        self._all_batch_indices: List[int] = []
+
+    @property
+    def done(self) -> bool:
+        """Ends prediction when the iteration count exceeds the total number of available batches"""
+        return self.iteration_count >= self._dl_max_batches
+
+    @property
+    def should_store_predictions(self) -> bool:
+        """whether the predictions should be stored for later usage (e.g. aggregation or returning)"""
+        any_pred = any(cb.interval.on_epoch for cb in self.trainer.prediction_writer_callbacks)
+        return self.return_predictions or any_pred
+
+    def reset(self) -> None:
+        """Resets the loops internal state"""
+        self.iteration_count = 0
+        self._all_batch_indices: List[int] = []
+        self.predictions: List[Any] = []
+
+    def on_run_start(
+        self,
+        dataloader_iter: Iterator,
+        dataloader_idx: int,
+        dl_max_batches: int,
+        num_dataloaders: int,
+        return_predictions: bool = False
+    ) -> None:
+        """Prepares the loops internal state
+
+        Args:
+            dataloader_iter: the iterator over the current dataloader
+            dataloader_idx: the index of the current dataloader
+            dl_max_batches: the maximum number of batches the current loader can produce
+            num_dataloaders: the total number of dataloaders
+            return_predictions: whether to return the obtained predictions
+        """
+        void(dataloader_iter, dataloader_idx)
```

Comment:
Should all these arguments be passed with keywords so they can be hidden with `kwargs`?

Otherwise, should we not include them in the docstring?

Reply:
@justusschock what was your intention here? could do either

Reply:
I think all the arguments that are not currently used can become **kwargs.

For removal of the docstrings: I have mixed feelings on removing them: on the one hand, they are currently unused, but on the other hand when we might open this part to user-API, I think it might be nice for them to know what the passed objects are.

I'm fine with either way.

Reply:
ok, I vote for keeping the args for now, we don't expose the docs yet in the api reference, so we may re-evaluate this when we do?

Reply:
>  I vote for keeping the args for now

I'm fine with this if they become kwargs

Reply:
I wasn't sure what you wanted me to do
https://github.com/PyTorchLightning/pytorch-lightning/pull/7700/commits/2639c784a0ff68ac8d8e649cacf8d8195990fb87

Reply:
I meant doing:

```python
    def on_run_start(
        self,
        dl_max_batches: int,
        num_dataloaders: int,
        return_predictions: bool = False,
        **kwargs: Any,  # dataloader_iter, dataloader_idx
    ) -> None:
```

Reply:
but `dataloader_iter, dataloader_idx` are supposed to be mandatory and documented.
I am highly confused by this, apologies.

Reply:
> are supposed to be mandatory and documented.

My point is that the fact that we "void" them suggests me that they should not be mandatory (aka passed as kwargs)

Reply:
the reviews recommended this, as these were unused variables and that's apparently bad. this is also why a call to void() was added. 

these args apart from `return_predictions` are definitely mandatory and do not have defaults at this moment.